### PR TITLE
Issue #SB-8263 feat: Actor layer

### PR DESCRIPTION
**Pull request by sudhirgiri2911**
Actor layer

### CODE REVIEW

File: actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.sunbird.common.models.util.JsonKey;
+import org.sunbird.common.models.util.LoggerEnum;
+import org.sunbird.common.models.util.ProjectLogger;
+import org.sunbird.common.models.util.PropertiesCache;
+
+public class ChannelUtil {
+  public static String getRootOrgDetailsUrl = null;
+
+  static {
+    String baseUrl = System.getenv(JsonKey.SUNBIRD_WEB_URL);
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java]: Use ProjectUtil.getConfigValue

File: actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    String searchPath = System.getenv(JsonKey.SUNBIRD_CHANNEL_API);
+    if (StringUtils.isBlank(searchPath))
+      searchPath = PropertiesCache.getInstance().getProperty(JsonKey.SUNBIRD_CHANNEL_API);
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java]: Use ProjectUtil.getConfigValue

File: actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import org.sunbird.common.models.util.LoggerEnum;
+import org.sunbird.common.models.util.ProjectLogger;
+import org.sunbird.common.models.util.PropertiesCache;
+
+public class ChannelUtil {
+  public static String getRootOrgDetailsUrl = null;
+
+  static {
+    String baseUrl = System.getenv(JsonKey.SUNBIRD_WEB_URL);
+    String searchPath = System.getenv(JsonKey.SUNBIRD_CHANNEL_API);
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java]: readChannelPath

File: actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      searchPath = PropertiesCache.getInstance().getProperty(JsonKey.SUNBIRD_CHANNEL_API);
+    getRootOrgDetailsUrl = baseUrl + searchPath;
+  }
+
+  private static Map<String, String> getHeaders(Map<String, String> headers) {
+    if (headers == null) {
+      headers = new HashMap<>();
+    }
+    headers.put(
+        HttpHeaders.AUTHORIZATION, JsonKey.BEARER + System.getenv(JsonKey.SUNBIRD_AUTHORIZATION));
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java]: Use ProjectUtil.getConfigValue

File: actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    return headers;
+  }
+
+  @SuppressWarnings("unchecked")
+  public static Map<String, Object> getRootOrgDetails(String rootOrgId) {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java]: readChannel(String channel)

File: actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    ObjectMapper mapper = new ObjectMapper();
+    Map<String, Object> resMap = new HashMap<>();
+    String response = "";
+    JSONObject jObject;
+    Map<String, String> headers = getHeaders(null);
+    ProjectLogger.log("making call to read RootOrg Details ==" + rootOrgId, LoggerEnum.INFO.name());
+    try {
+      response = HttpUtil.sendGetRequest(getRootOrgDetailsUrl + "/" + rootOrgId, headers);
+      ProjectLogger.log("RootOrg Read details are ==" + response, LoggerEnum.INFO.name());
+      jObject = new JSONObject(response);
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java]: Use ObjectMapper

File: actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      String apiId = jObject.getString("id");
+      String resmsgId = (String) jObject.getJSONObject("params").get("resmsgid");
+      String status = (String) jObject.getJSONObject(JsonKey.PARAMS).get(JsonKey.STATUS);
+      String resultStr = jObject.getString(JsonKey.RESULT);
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java]: Check status is successful or not.

File: actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      Map<String, Object> data = mapper.readValue(resultStr, Map.class);
+      Map<String, Object> param = new HashMap<>();
+      param.put(JsonKey.RES_MSG_ID, resmsgId);
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ChannelUtil.java]: Remove unnecessary fields.

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 9]
```diff
@@ -28,6 +28,10 @@
   private static Map<String, Object> roleMap = new ConcurrentHashMap<>();
   private static Map<String, String> orgTypeMap = new ConcurrentHashMap<>();
   private static Map<String, String> configSettings = new ConcurrentHashMap<>();
+  private static Map<String, Map<String, List<Map<String, String>>>> frameworkMap =
+      new ConcurrentHashMap<>();
+  private static Map<String, List<String>> frameworkFieldsConfig = new ConcurrentHashMap<>();
+  private static Map<String, String> channelFrameworkMap = new ConcurrentHashMap<>();
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: readChannelResponseMap
readFrameworkResponseMap

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 10]
```diff
@@ -66,6 +71,16 @@ private void cacheSystemConfig(Map<String, String> configSettings) {
     }
   }
 
+  private void cacheFrameworkFieldsConfig() {
+    Response response =
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: Use SystemSettingClient

File: actors/common/src/main/java/org/sunbird/learner/util/FrameworkUtil.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+import org.sunbird.common.models.util.HttpUtil;
+import org.sunbird.common.models.util.JsonKey;
+import org.sunbird.common.models.util.LoggerEnum;
+import org.sunbird.common.models.util.ProjectLogger;
+import org.sunbird.common.models.util.PropertiesCache;
+import org.sunbird.common.models.util.StringFormatter;
+import org.sunbird.common.responsecode.ResponseCode;
+import org.sunbird.models.user.User;
+
+public class FrameworkUtil {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/FrameworkUtil.java]: ChannelUtil and FrameworkUtil => ContentStoreUtil

File: actors/common/src/main/java/org/sunbird/learner/util/FrameworkUtil.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+      param.put(JsonKey.STATUS, status);
+      resMap.put(JsonKey.PARAMS, param);
+      resMap.put(JsonKey.RESULT, data);
+    } catch (IOException | JSONException e) {
+      ProjectLogger.log("Error found during contnet search parse==" + e.getMessage(), e);
+    }
+    return resMap;
+  }
+
+  public static void validateFrameworkUpdateRequest(
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/FrameworkUtil.java]: This method should be in UserRequestValidator

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java (hunk 10 above)
Comment (sudhirgiri2911) [reply actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: SystemSettingClientImpl.getSystemSettingByFieldAndKey() it takes ( ActorRef actorRef) as an argument, how to pass it through DataCacheHandler.

File: actors/common/src/main/java/org/sunbird/learner/util/ContentStoreUtil.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.http.HttpHeaders;
+import org.sunbird.common.models.util.HttpUtil;
+import org.sunbird.common.models.util.JsonKey;
+import org.sunbird.common.models.util.LoggerEnum;
+import org.sunbird.common.models.util.ProjectLogger;
+import org.sunbird.common.models.util.ProjectUtil;
+
+public class ContentStoreUtil {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ContentStoreUtil.java]: Move this to sunbird-utils/content-util (new project).

content-util/src/main/java/org/sunbird/content/util/ContentStoreUtil.java

File: actors/common/src/main/java/org/sunbird/learner/util/ContentStoreUtil.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+  private static Map<String, String> getHeaders() {
+    Map<String, String> headers = new HashMap<>();
+    headers.put(
+        HttpHeaders.AUTHORIZATION, JsonKey.BEARER + System.getenv(JsonKey.CONTENT_STORE_API_KEY));
+    return headers;
+  }
+
+  @SuppressWarnings("unchecked")
+  public static Map<String, Object> getReadDetails(String id, String readPath) {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ContentStoreUtil.java]: getChannelDetails
getFrameworkDetails

File: actors/common/src/main/java/org/sunbird/learner/util/ContentStoreUtil.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+      String requestUrl =
+          ProjectUtil.getConfigValue(JsonKey.SUNBIRD_STORE_API_BASE_URL)
+              + ProjectUtil.getConfigValue(readPath)
+              + "/"
+              + id;
+      String response = HttpUtil.sendGetRequest(requestUrl, headers);
+      ProjectLogger.log("Read details are ==" + response, LoggerEnum.INFO.name());
+      resultMap = mapper.readValue(response, Map.class);
+      if (!((String) resultMap.get(JsonKey.RESPONSE_CODE)).equalsIgnoreCase(JsonKey.OK)) {
+        ProjectLogger.log("ContentStoreUtil : GetReadDetails Read Error Obtained");
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/ContentStoreUtil.java]: Return null if not OK.

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java (hunk 10 above)
Comment (bvinayakumar) [reply actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: Framework fields config need not be stored in DataCacheHandler. Instead save it in the user management actor.

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 16]
```diff
@@ -975,4 +986,64 @@ private void setDefaultUserProfileVisibility(Map<String, Object> userMap) {
...
   }
+
+  @SuppressWarnings("unchecked")
+  public static String getFrameworkId(String channel) {
+    String frameworkId = DataCacheHandler.getchannelFrameworkMap().get(channel);
+    if (frameworkId != null) {
+      Map<String, Object> resultMap =
+          ContentStoreUtil.getReadDetails(channel, JsonKey.SUNBIRD_CHANNEL_READ_API);
+      Map<String, Object> results = (Map<String, Object>) resultMap.get(JsonKey.RESULT);
+      Map<String, Object> channelDetails = (Map<String, Object>) results.get(JsonKey.CHANNEL);
```
Comment (bvinayakumar) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: Add null check on results, channelDetails, frameworkId before updating channel framework ID map.

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 17]
```diff
@@ -28,6 +30,10 @@
   private static Map<String, Object> roleMap = new ConcurrentHashMap<>();
   private static Map<String, String> orgTypeMap = new ConcurrentHashMap<>();
   private static Map<String, String> configSettings = new ConcurrentHashMap<>();
+  private static Map<String, Map<String, List<Map<String, String>>>> readFrameworkResponseMap =
+      new ConcurrentHashMap<>();
+  private static Map<String, List<String>> frameworkFieldsConfig = new ConcurrentHashMap<>();
+  private static Map<String, String> readChannelResponseMap = new HashMap<>();
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: readChannelResponseMap -> channelFrameworkIdMap

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 18]
```diff
@@ -174,4 +204,41 @@ public static void setOrgTypeMap(Map<String, String> orgTypeMap) {
...
+    DataCacheHandler.readFrameworkResponseMap.put(frameworkId, frameworkCacheMap);
+  }
+
+  /** @param channelFrameworkMap the channelFrameworkMap to set */
+  public static void setChannelFrameworkMap(Map<String, String> channelFrameworkMap) {
+    DataCacheHandler.readChannelResponseMap = channelFrameworkMap;
+  }
+
+  /** @return the channelFrameworkMap */
+  public static Map<String, String> getchannelFrameworkMap() {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: getChannelFrameworkIdMap

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 19]
```diff
@@ -174,4 +204,41 @@ public static void setOrgTypeMap(Map<String, String> orgTypeMap) {
...
+    return readChannelResponseMap;
+  }
+
+  /** @param channelFrameworkMap the channelFrameworkMap to set */
+  public static void updateChannelFrameworkMap(String channel, String frameworkId) {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: updateChannelFrameworkIdMap

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 20]
```diff
@@ -975,4 +986,64 @@ private void setDefaultUserProfileVisibility(Map<String, Object> userMap) {
...
+  private static void getFrameworkDataAndCache(String frameworkId) {
+    Map<String, Object> response =
+        ContentStoreUtil.getReadDetails(frameworkId, JsonKey.SUNBIRD_FRAMEWORK_READ_API);
+    Map<String, List<Map<String, String>>> frameworkCacheMap = new HashMap<>();
+    List<String> supportedfFields = DataCacheHandler.getFrameworkFieldsConfig().get(JsonKey.FIELDS);
+    Map<String, Object> result = (Map<String, Object>) response.get(JsonKey.RESULT);
+    Map<String, Object> frameworkDetails = (Map<String, Object>) result.get(JsonKey.FRAMEWORK);
+    List<Map<String, Object>> frameworkCategories =
+        (List<Map<String, Object>>) frameworkDetails.get(JsonKey.CATEGORIES);
+    for (Map<String, Object> frameworkCategoriesValue : frameworkCategories) {
```
Comment (bvinayakumar) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: Add null check on result, frameworkDetails and frameworkCategories

Comment (bvinayakumar) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: frameworkCategoriesValue -> frameworkCategoryValues

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 21]
```diff
@@ -975,4 +986,64 @@ private void setDefaultUserProfileVisibility(Map<String, Object> userMap) {
...
+      String frameworkField = (String) frameworkCategoriesValue.get(JsonKey.CODE);
+      List<Map<String, String>> listOfFields = new ArrayList<>();
+      if (supportedfFields.contains(frameworkField)
+          || frameworkField.equalsIgnoreCase(JsonKey.GRADE_LEVEL)) {
```
Comment (bvinayakumar) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: Don't have any special check for GRADE_LEVEL. Update configuration instead.

e.g. userProfileConfig.mandatoryFields = ["board", "gradeLevel", "medium"]

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 22]
```diff
@@ -975,4 +986,64 @@ private void setDefaultUserProfileVisibility(Map<String, Object> userMap) {
...
+        if (frameworkField.equalsIgnoreCase(JsonKey.GRADE_LEVEL)) frameworkField = JsonKey.CLASS;
+        List<Map<String, Object>> frameworkTerms =
```
Comment (bvinayakumar) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: Add null check on frameworkTerms.

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 23]
```diff
@@ -975,4 +986,64 @@ private void setDefaultUserProfileVisibility(Map<String, Object> userMap) {
...
+            (List<Map<String, Object>>) frameworkCategoriesValue.get(JsonKey.TERMS);
+        for (Map<String, Object> frameworkTermsField : frameworkTerms) {
```
Comment (bvinayakumar) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: frameworkTermsField -> frameworkTerm

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java (hunk 22 above)
Comment (bvinayakumar) [reply actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: frameworkTerms -> frameworkTermList

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 24]
```diff
@@ -28,6 +30,10 @@
   private static Map<String, Object> roleMap = new ConcurrentHashMap<>();
   private static Map<String, String> orgTypeMap = new ConcurrentHashMap<>();
   private static Map<String, String> configSettings = new ConcurrentHashMap<>();
+  private static Map<String, Map<String, List<Map<String, String>>>> readFrameworkResponseMap =
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: readFrameworkResponseMap -> frameworkCategoriesMap

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 25]
```diff
@@ -174,4 +204,41 @@ public static void setOrgTypeMap(Map<String, String> orgTypeMap) {
...
     DataCacheHandler.configSettings = configSettings;
   }
+
+  /** @return the frameworkMap */
+  public static Map<String, Map<String, List<Map<String, String>>>> getFrameworkMap() {
+    return readFrameworkResponseMap;
+  }
+
+  /** @param frameworkMap the frameworkValues to set */
+  public static void setFrameworkMap(
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: setFrameworkCategoriesMap

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 26]
```diff
@@ -174,4 +204,41 @@ public static void setOrgTypeMap(Map<String, String> orgTypeMap) {
   public static void setConfigSettings(Map<String, String> configSettings) {
     DataCacheHandler.configSettings = configSettings;
   }
+
+  /** @return the frameworkMap */
+  public static Map<String, Map<String, List<Map<String, String>>>> getFrameworkMap() {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: getFrameworkCategoriesMap

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 27]
```diff
@@ -174,4 +204,41 @@ public static void setOrgTypeMap(Map<String, String> orgTypeMap) {
...
+  }
+
+  /** @param frameworkId and its Values and update it */
+  public static void updateFrameworkMap(
+      String frameworkId, Map<String, List<Map<String, String>>> frameworkCacheMap) {
+    DataCacheHandler.readFrameworkResponseMap.put(frameworkId, frameworkCacheMap);
+  }
+
+  /** @param channelFrameworkMap the channelFrameworkMap to set */
+  public static void setChannelFrameworkMap(Map<String, String> channelFrameworkMap) {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: setChannelFrameworkIdMap

File: actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java [hunk 28]
```diff
@@ -174,4 +204,41 @@ public static void setOrgTypeMap(Map<String, String> orgTypeMap) {
...
+    DataCacheHandler.readFrameworkResponseMap = frameworkMap;
+  }
+
+  /** @return the frameworkFieldsConfig */
+  public static Map<String, List<String>> getFrameworkFieldsConfig() {
+    return frameworkFieldsConfig;
+  }
+
+  /** @param frameworkId and its Values and update it */
+  public static void updateFrameworkMap(
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/util/DataCacheHandler.java]: updateFrameworkCategoriesMap

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 29]
```diff
@@ -688,9 +689,20 @@ private void updateUser(Request actorMessage) {
     Util.getUserProfileConfig(systemSettingActorRef);
     userService.validateUserId(actorMessage);
     Map<String, Object> userMap = actorMessage.getRequest();
+    List<String> frameworkFields = DataCacheHandler.getFrameworkFieldsConfig().get(JsonKey.FIELDS);
+    List<String> frameworkMandatoryFields =
+        DataCacheHandler.getFrameworkFieldsConfig().get(JsonKey.MANDATORY_FIELDS);
     userRequestValidator.validateUpdateUserRequest(actorMessage);
+    userRequestValidator.validateFrameworkRequestData(
```
Comment (bvinayakumar) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: validateFrameworkRequestData -> validateMandatoryFrameworkFields

File: actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java [hunk 30]
```diff
@@ -688,9 +689,20 @@ private void updateUser(Request actorMessage) {
...
+        userMap, frameworkFields, frameworkMandatoryFields);
     Map<String, Object> userDbRecord = UserUtil.validateExternalIdsAndReturnActiveUser(userMap);
-
+    if (userMap.containsKey(JsonKey.FRAMEWORK)) {
+      User userObj = userService.getUserById((String) userMap.get(JsonKey.USER_ID));
+      String frameworkId = getFrameworkId(userObj.getChannel());
+      Map<String, List<Map<String, String>>> frameworkCachedValue =
+          getFrameworkDetails(frameworkId);
+      userRequestValidator.validateFrameworkRequestDetails(userMap, frameworkCachedValue);
```
Comment (bvinayakumar) [inline actors/user/src/main/java/org/sunbird/user/actors/UserManagementActor.java]: validateFrameworkRequestDetails -> validateFrameworkCategoryValues

**Pull request merged**